File: example/step-definitions/desktop.steps.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { ConductorWorld } from 'conductor';
import * as path from 'path';

const DESKTOP_TIMEOUT = { timeout: 60000 };

interface TodoResponse {
  id: number;
  title: string;
  status: string;
  priority: string;
}

async function ensureApi(world: ConductorWorld): Promise<void> {
  if (!world.api.isInitialized) await world.api.init();
}

async function findTodoIdByTitle(world: ConductorWorld, title: string): Promise<number> {
  await ensureApi(world);
  const response = await world.api.get(`${world.config.api.baseUrl}/todos`);
  const todos = await response.json() as TodoResponse[];
  const todo = todos.find(t => t.title === title);
  if (!todo) throw new Error(`Todo "${title}" not found via API`);
  return todo.id;
}

Given('the desktop app is running', DESKTOP_TIMEOUT, async function (this: ConductorWorld) {
  const jarPath = path.resolve('../apps/desktop/build/libs/todoapp-desktop-all.jar');
  const apiUrl = process.env.API_BASE_URL_FOR_DESKTOP ?? this.config.api.baseUrl;
  this.logger.info({ jarPath, apiUrl, agentJar: this.config.desktop?.agentJar }, 'Launching desktop app');
  try {
    await this.fx.launch({
      app: 'com.example.todoapp.Launcher',
      classpath: jarPath,
      jvmArgs: [`-DAPI_BASE_URL=${apiUrl}`],
      readyTimeoutMs: 30000,
    });
  } catch (e: any) {
    this.logger.error({ error: e.message, stack: e.stack, details: e.details }, 'Desktop launch failed');
    throw e;
  }
  await this.fx.locator('#todo-input').waitFor({ state: 'visible', timeout: 15000 });
});

When('I create a todo {string} with priority {string} via the desktop app', DESKTOP_TIMEOUT, async function (this: ConductorWorld, title: string, priority: string) {
  await this.fx.locator('#todo-input').fill(title);
  await this.fx.locator('#todo-priority-select').selectOption(priority);
  await this.fx.locator('#todo-add').click();
  await this.fx.locator(`text=${title}`).waitFor({ state: 'visible' });
  this.data.lastTodoTitle = title;
});

When('I create a todo {string} via the desktop app', DESKTOP_TIMEOUT, async function (this: ConductorWorld, title: string) {
  await this.fx.locator('#todo-input').fill(title);
  await this.fx.locator('#todo-add').click();
  await this.fx.locator(`text=${title}`).waitFor({ state: 'visible' });
  this.data.lastTodoTitle = title;
});

async function refreshDesktop(world: ConductorWorld): Promise<void> {
  await world.fx.locator('#refresh-btn').click();
}

When('I edit the todo {string} to {string} via the desktop app', DESKTOP_TIMEOUT, async function (this: ConductorWorld, currentTitle: string, newTitle: string) {
  const id = await findTodoIdByTitle(this, currentTitle);
  await refreshDesktop(this);
  await this.fx.locator(`#edit-${id}`).waitFor({ state: 'visible', timeout: 10000 });
  await this.fx.locator(`#edit-${id}`).click();
  await this.fx.locator('text=OK').waitFor({ state: 'visible', timeout: 5000 });
  // Dialog's text field uses .text-input style class — query all and pick the focused one (dialog field)
  await this.fx.locator('css=.dialog-pane .text-input').setText(newTitle);
  await this.fx.locator('text=OK').click();
  await this.fx.locator(`text=${newTitle}`).waitFor({ state: 'visible' });
  await new Promise(r => setTimeout(r, 500));
});

When('I toggle the todo {string} via the desktop app', DESKTOP_TIMEOUT, async function (this: ConductorWorld, title: string) {
  const id = await findTodoIdByTitle(this, title);
  await refreshDesktop(this);
  await this.fx.locator(`#toggle-${id}`).waitFor({ state: 'visible', timeout: 10000 });
  await this.fx.locator(`#toggle-${id}`).click();
  // Wait for the async PUT to complete server-side
  await new Promise(r => setTimeout(r, 800));
});

When('I delete the todo {string} via the desktop app', DESKTOP_TIMEOUT, async function (this: ConductorWorld, title: string) {
  const id = await findTodoIdByTitle(this, title);
  await refreshDesktop(this);
  await this.fx.locator(`#delete-${id}`).waitFor({ state: 'visible', timeout: 10000 });
  await this.fx.locator(`#delete-${id}`).click();
  await this.fx.locator('text=OK').waitFor({ state: 'visible', timeout: 5000 });
  await this.fx.locator('text=OK').click();
});

Then('the todo {string} should be visible on the desktop app', DESKTOP_TIMEOUT, async function (this: ConductorWorld, title: string) {
  await this.fx.locator(`text=${title}`).waitFor({ state: 'visible' });
});

Then('the todo {string} should not be visible on the desktop app', DESKTOP_TIMEOUT, async function (this: ConductorWorld, title: string) {
  await this.fx.locator(`text=${title}`).waitFor({ state: 'hidden', timeout: 5000 });
});

Then('I take a desktop screenshot {string}', DESKTOP_TIMEOUT, async function (this: ConductorWorld, name: string) {
  const slug = name.replace(/\s+/g, '-').toLowerCase();
  try {
    const buffer = await this.fx.screenshot(slug);
    await this.attach(buffer, 'image/png');
  } catch (e: any) {
    this.logger.warn({ error: e.message }, 'Desktop screenshot failed');
  }
});

File: java/conductor-example/src/test/java/com/nouhouari/conductor/example/stepdefs/DesktopSteps.java
package com.nouhouari.conductor.example.stepdefs;

import com.nouhouari.conductor.drivers.JavaFxDriver;
import com.nouhouari.conductor.world.ConductorWorld;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import java.nio.file.Paths;
import java.util.List;

/** Java port of example/step-definitions/desktop.steps.ts (JavaFX, via JavaFxDriver / fxagent.jar). */
public class DesktopSteps {

    private final ConductorWorld world;

    public DesktopSteps(ConductorWorld world) {
        this.world = world;
    }

    @Given("the desktop app is running")
    public void theDesktopAppIsRunning() {
        String jarPath = Paths.get("../apps/desktop/build/libs/todoapp-desktop-all.jar").toAbsolutePath().toString();
        String apiUrl = System.getenv("API_BASE_URL_FOR_DESKTOP") != null
                ? System.getenv("API_BASE_URL_FOR_DESKTOP")
                : world.config.api().baseUrl();
        world.logger.info("Launching desktop app: jarPath={} apiUrl={} agentJar={}",
                jarPath, apiUrl, world.config.desktop() != null ? world.config.desktop().agentJar() : null);
        try {
            world.fx().launch(new JavaFxDriver.LaunchOptions(
                    "com.example.todoapp.Launcher",
                    jarPath,
                    List.of("-DAPI_BASE_URL=" + apiUrl),
                    30000));
        } catch (RuntimeException e) {
            world.logger.error("Desktop launch failed: {}", e.getMessage());
            throw e;
        }
        world.fx().locator("#todo-input").waitFor(new JavaFxDriver.WaitOptions("visible", 15000));
    }

    @When("I create a todo {string} with priority {string} via the desktop app")
    public void iCreateATodoWithPriorityViaTheDesktopApp(String title, String priority) {
        world.fx().locator("#todo-input").fill(title);
        world.fx().locator("#todo-priority-select").selectOption(priority);
        world.fx().locator("#todo-add").click();
        world.fx().locator("text=" + title).waitFor(new JavaFxDriver.WaitOptions("visible", null));
        world.data.put("lastTodoTitle", title);
    }

    @When("I create a todo {string} via the desktop app")
    public void iCreateATodoViaTheDesktopApp(String title) {
        world.fx().locator("#todo-input").fill(title);
        world.fx().locator("#todo-add").click();
        world.fx().locator("text=" + title).waitFor(new JavaFxDriver.WaitOptions("visible", null));
        world.data.put("lastTodoTitle", title);
    }

    private void refreshDesktop() {
        world.fx().locator("#refresh-btn").click();
    }

    @When("I edit the todo {string} to {string} via the desktop app")
    public void iEditTheTodoToViaTheDesktopApp(String currentTitle, String newTitle) throws InterruptedException {
        int id = ApiHelpers.findTodoIdByTitle(world, currentTitle);
        refreshDesktop();
        world.fx().locator("#edit-" + id).waitFor(new JavaFxDriver.WaitOptions("visible", 10000));
        world.fx().locator("#edit-" + id).click();
        world.fx().locator("text=OK").waitFor(new JavaFxDriver.WaitOptions("visible", 5000));
        // Dialog's text field uses .text-input style class — query all and pick the focused one (dialog field)
        world.fx().locator("css=.dialog-pane .text-input").setText(newTitle);
        world.fx().locator("text=OK").click();
        world.fx().locator("text=" + newTitle).waitFor(new JavaFxDriver.WaitOptions("visible", null));
        Thread.sleep(500);
    }

    @When("I toggle the todo {string} via the desktop app")
    public void iToggleTheTodoViaTheDesktopApp(String title) throws InterruptedException {
        int id = ApiHelpers.findTodoIdByTitle(world, title);
        refreshDesktop();
        world.fx().locator("#toggle-" + id).waitFor(new JavaFxDriver.WaitOptions("visible", 10000));
        world.fx().locator("#toggle-" + id).click();
        // Wait for the async PUT to complete server-side
        Thread.sleep(800);
    }

    @When("I delete the todo {string} via the desktop app")
    public void iDeleteTheTodoViaTheDesktopApp(String title) {
        int id = ApiHelpers.findTodoIdByTitle(world, title);
        refreshDesktop();
        world.fx().locator("#delete-" + id).waitFor(new JavaFxDriver.WaitOptions("visible", 10000));
        world.fx().locator("#delete-" + id).click();
        world.fx().locator("text=OK").waitFor(new JavaFxDriver.WaitOptions("visible", 5000));
        world.fx().locator("text=OK").click();
    }

    @Then("the todo {string} should be visible on the desktop app")
    public void theTodoShouldBeVisibleOnTheDesktopApp(String title) {
        world.fx().locator("text=" + title).waitFor(new JavaFxDriver.WaitOptions("visible", null));
    }

    @Then("the todo {string} should not be visible on the desktop app")
    public void theTodoShouldNotBeVisibleOnTheDesktopApp(String title) {
        world.fx().locator("text=" + title).waitFor(new JavaFxDriver.WaitOptions("hidden", 5000));
    }

    @Then("I take a desktop screenshot {string}")
    public void iTakeADesktopScreenshot(String name) {
        String slug = name.replaceAll("\\s+", "-").toLowerCase();
        try {
            byte[] buffer = world.fx().screenshot(slug);
            world.scenario.attach(buffer, "image/png", slug);
        } catch (RuntimeException e) {
            world.logger.warn("Desktop screenshot failed: {}", e.getMessage());
        }
    }
}
